GEP-1897 Explicit Backend TLS Connection Configuration (was TLS from Gateway to Backend...) Update with API details (#2113)
kubernetes-sigs/gateway-api: geps/gep-1897.md

a backend pod that has its own certificate”. TLS configuration can be a nebulous topic,
so in order to drive resolution this GEP focuses only on this single piece of functionality.

Furthermore, for Gateway API to handle the case where the service or backend owner is doing their own TLS, _and_
the service or backend owner wants to validate the clients connecting to it, two things need to happen:

> - The service or backend owner has to provide a method for the Gateway owner to retrieve a certificate.

Author: We're not addressing the mTLS use case in this proposal.

> 1. The solution must satisfy the following use case: the backend pod has its own

Reviewer: I was thinking of a use-case where a backend (pod/service) is also receiving mesh traffic (in addition to traffic from the gateway) and is configured to only accept mTLS connections. Can we enhance the `TLSConnectionPolicy` struct to add:

- a mode field (to specify mTLS or TLS)
- a field similar to `TrustedCACertRefs` - say `ClientCertRefs` - which the client will use to provide client cert in case of mTLS

So that this API can be directly used for the mesh-case and also cases where the Gateway could or would like to provide a client-cert with an mTLS connection?

Reviewer 2: I think mTLS is intentionally out of scope for this phase of the GEP, but we want to ensure we have room to add it in the future. So it's probably worth sketching out some possible future extensions for mTLS here, but I don't want to block on deciding on a solution here, just as long as we agree that we've left room for a future solution.

Author: The mesh case is also intentionally out of scope for this tightly-scoped proposal.  In the case we'd need to do something for mTLS, I would rather propose a new `mTLSConnectionPolicy`.

Reviewer: I guess a new mTLSConnectionPolicy type does not sound that bad after all although a single type like TLSConnectionPolicy supporting mTLS is preferable in which case we should make sure it's possible to add it later.
 
I disagree that mTLS is not a service level concern so should be left out. mTLS = TLS + client certificate. If TLS is a service level concern, then so is mTLS. Also mTLS provides client identity which is used in authz and other(?) mesh configs.

Reviewer 3: I think the issue here is one of term overloading.

@sanjaypujare appears to be using `mTLS` to mean "a connection between a Gateway and backend where the Gateway verifies the backend and the backend verifies the Gateway", or in other words a mutual TLS handshake where the certificate management is done by _the owner of the service_, whereas I think @keithmattix and @howardjohn are more thinking about the Service Mesh use of `mTLS` to mean "a mutual auth handshake where the certificate management is performed silently and transparently by infrastructure". In the latter case, it's _important_ that Gateway API constructs are not required, because if they are, then the certificate management is no longer transparent.

The important part here is that this GEP is _not_ about magic infrastructure that automatically upgrades connections to TLS on your behalf, it's about an app owner being able to say that their app _expects_ TLS and _has no magic to make it work transparently_.

This is why, in the Listener case, I've pushed very hard for us to call similar functionality "client certificate configuration" rather than "mTLS", because the use of the acronym can sometimes, depending on who's using it, carry a _lot_ of extra meaning that others may not know is there.

Now, can I see a use for Client Certificate settings in the future in TLSConnectionPolicy? Possibly, but I think it's important that we keep in mind personas here. The whole idea of having mutual authentication of certificates is that it's expected that the two parties spoke _independently_ to the Certificate Authority, and took some steps to prove that they are who they say they are, and then the Certificate Authority provides a method for the two parties to check that the CA that they both trust says the other is okay.

Having the service or backend owner provide the _client certificate_ details completely bypasses this - at that point, there is very little gain in using TLS at all.

For Gateway API to handle the "service/backend owner is doing their own TLS", _and_ "service or backend owner wants to validate the clients connecting to it", it's on the service or backend owner to provide a method for the Gateway owner to retrieve a certificate, and on _us_, as Gateway API designers, to provide a place _on the Gateway_ to configure the client _keypair_ that should be used (since it's both a private and a public key that are required in the client certificate case).

So, what's the summary of this rambling diatribe?
- We've excluded "mTLS" from this precisely because the fact that different people can use it to mean different things means we end up having discussions _exactly like this one, over and over again_.
- This GEP is about _explicit_ TLS configuration, not the _implicit_, _automatic_ TLS configuration you get with a Service Mesh or similar infrastructure tooling. Perhaps we should update some of the preamble to this effect if this makes sense to people other than me, and we can avoid yet _another_ round of this discussion.
- Using an _explicitly configured client certificate_ for the Gateway end of a TLS connection is a reasonable use case, but that _should be configured on the Gateway_, not as part of the TLSConnectionPolicy.

Author: Nicely put.  I'll add some of this to the preamble.

## Longer Term Goals

These are worthy goals, but may need a different GEP for proper attention.  This GEP is concerned entirely with the
controlplane, i.e. the hop between gateway and backend.

1. [TCPRoute](https://gateway-api.sigs.k8s.io/references/spec/#gateway.networking.k8s.io/v1alpha2.TCPRoute) and
[GRPCRoute](https://gateway-api.sigs.k8s.io/references/spec/#gateway.networking.k8s.io/v1alpha2.GRPCRoute) use cases
are not addressed here, because at this point in time these two route types are not graduated to beta.
2. Mutual TLS (mTLS) use cases are not thoroughly addressed here because we have tried and failed in the past to
generalize around mTLS, and the treatment of mTLS should be taken outside the context of the hop between gateway and

> backend.

Author: Okay, I will incorporate this.

> backend.

Author: Actually, based on the commentary from Nick, I'm going to use this.  It doesn't include what might need to happen in the future for mTLS.  Sound okay?

Mutual TLS (mTLS) use cases are intentionally out of scope for this GEP because this GEP does not address the case where connections to TLS are implicitly configured on behalf of the user, such as in Service Mesh.  It is about the case where an application developer needs to explicitly express that they expect TLS when there is no automatic, implicit configuration of that.

> backend.

Author: Was updated again later.

3. Service mesh use cases are not addressed here because the main concern is the controlplane hop between gateway

> and backend, which is only one part of the picture in a service mesh.

Author: I couldn't have said it better myself.  Will do.

## Longer Term Goals

These are worthy goals, but may need a different GEP for proper attention.  This GEP is concerned entirely with the
controlplane, i.e. the hop between gateway and backend.

1. [TCPRoute](https://gateway-api.sigs.k8s.io/references/spec/#gateway.networking.k8s.io/v1alpha2.TCPRoute) and
[GRPCRoute](https://gateway-api.sigs.k8s.io/references/spec/#gateway.networking.k8s.io/v1alpha2.GRPCRoute) use cases
are not addressed here, because at this point in time these two route types are not graduated to beta.

> 2. Mutual TLS (mTLS) use cases are intentionally out of scope for this GEP because this GEP does not address the case

Reviewer 4: Sorry to keep hammering on this point, but I am not sure this makes sense and it was my only pushback on this GEP so I think its important to get it right.

**Mutual TLS being out of scope is not because of mesh.** 

>  It is about the  case where an application developer needs to explicitly express that they expect TLS when there is no automatic, implicit configuration of that.

Its perfectly valid for an implementation to expect mTLS explicitly. mTLS is not even really an independent thing, its just TLS with a field set - its a name for a subset of TLS.

Mutual TLS is not addressed by the GEP because _the API fundamentally cannot support mutual authentication_ because of its design of being backend-attached.

tl;dr

* Not supporting "mesh transport" is an explicit design goal
* Not supporting "mutual TLS" is a down-scoping of the GEP to make sure we can get something merged. However, the design is not future proofed, so despite this being a longer term goal it cannot be extended
* Mutual TLS != Mesh Transport

Author: My goal was to highlight mesh as an example of the "case where connections to TLS are implicitly configured on the behalf of the user".  I will try to add that the "API cannot support mutual authentication because its design is backend-attached".

Choosing any other role would move the application-related responsibility from the application developer role to that
role, which violates the role-oriented design principle of Gateway API. As mentioned in Non-goal #7, providing a
mechanism for the cluster operator gateway role to override gateway to backend TLS settings is not covered here, but can
be addressed in a future update should the need arise.

TLSConnectionPolicy is defined as a Direct Policy Attachment without defaults or overrides, applied to a Service that
accesses the backend in question, where the TLSConnectionPolicy resides in the same namespace as the Service it is
applied to.  The TLSConnectionPolicy and the Service must reside in the same namespace in order to prevent the
complications involved with sharing trust across namespace boundaries.  By choosing the Service resource rather than the
Route resource, we can reuse the same TLSConnectionPolicy for all the different Routes that might point to this Service.
For the use case where certificates are stored in their own namespace, users may create Secrets and use ReferenceGrants
for a TLSConnectionPolicy-to-Secret binding.

> In the API defined here, the definition of TrustedCACertRefs follows a convention established by TLSRoute in

Reviewer 2: This links to a field on Gateway, so I think this may be more accurate:
```suggestion
In the API defined here, the definition of TrustedCACertRefs follows a convention established by the CertificateRefs field on Gateway in
```

> https://github.com/kubernetes-sigs/gateway-api/blob/18e79909f7310aafc625ba7c862dfcc67b385250/apis/v1beta1/gateway_types.go#L340

Reviewer 2: Can you turn this into a link? Our docs generation does not do that automatically.

backend destination certificate authority.  This solution proposes, as introduced in
[GEP-713](https://gateway-api.sigs.k8s.io/geps/gep-713/), that the implementation
should watch the connections to a specific port on a specified targetRef (such as a Service), and if the port and
Service match a TLSConnectionPolicy, then assume the connection is TLS, and verify that the targetRef’s certificate can
be validated by the provided trusted CA certificates before the connection is made.  On the question of how to signal
that there was a failure in the certificate validation, this is left up to the implementation to return a response error
that is appropriate, such as one of the HTTP error codes: 400 (Bad Request), 401 (Unauthorized), 403 (Forbidden), or
other signal that makes the failure sufficiently clear to the requester without revealing too much about the transaction,
based on established security requirements.

Not covered here, but possible to add would be additional configuration options mentioned in [SIG-NET Gateway API: TLS
to the K8s.Service/Backend](https://docs.google.com/document/d/1RTYh2brg_vLX9o3pTcrWxtZSsf8Y5NQvIG52lpFcZlo).  All of

> these are currently implementation-dependent, with the following recommended defaults:
> - Server Name Indication: enables passing of the server name through server name indication, in the TLS transaction, to

Reviewer 2: Our docs generation requires an empty line before the start of a list (https://deploy-preview-2113--kubernetes-sigs-gateway-api.netlify.app/geps/gep-1897/).

```suggestion
these are currently implementation-dependent, with the following recommended defaults:

- Server Name Indication: enables passing of the server name through server name indication, in the TLS transaction, to
```

assist with selection of certificates when several hosts share the same IP address. (default to enabled)

> - Subject Alternative Name certificates: enable the use of a single certificate that can serve multiple domains.

Reviewer 4: What is "enabled"? this is a property of the verification. So does enabled mean we should verify it (we ought to by default or it's the equivalent of -k in curl).

If so what are we verifying? It's not just a bool.

Browsers handle this by asserting the DNS name matches the certificate, but a Kubernetes service doesn't really have as strict of a mapping.

Reviewer: I think what "enabled" means is that a SAN (multi-domain) certificate will be accepted when true. When false, a certificate containing SANs will be rejected right away. The verification is always done (unless disabled otherwise with a different setting).

Regarding what we are verifying it's a good question: in Kubernetes (or service meshes like Istio) certificates may contain SPIFFE ids but the client needs to be provided with a mapping (service to valid SPIFFE id) to authorize the server.

Reviewer 4: SAN is the only non-deprecated way to encode a DNS name as far as I know (CN was deprecated 20 years ago). So it's not a question of if we verify San or not - we MUST.

The question is *what* SAN do we verify. Tou don't have to get into spife even, the pod could have any certificate..

Reviewer 4: the simplest case would be to assume svc.ns.svc.cluster.local but I am not sure that is sufficient, or even a common usage

Reviewer: I agree that SAN certs have been around for long and CN-only certs have been deprecated so this setting is not needed/useful.

Regarding "what SAN do we verify": may be that needs to be an additional field here with defaults for inferred names like "svc.ns.svc.cluster.local"

Author: My intention for "default is enabled" is that the option to use each of these is on by default.  Implementations could choose whether to allow any option to be disabled.  If that doesn't apply to SAN certificates, and you expect that no one would ever want to program "no SAN certificates allowed", then I can remove this and add a section that SAN certificates are always allowed.

Reviewer 4: "Using SAN" is not a meaningful statement.. it's not our choice,it's the TLS spec. We need to specify *which* SAN is allowed. This is absolutely a hard requirement for anything doing TLS

Author: Your feedback is not clear to me, can you help me out?  What do you expect to need to do with the SAN values other than provide them to the target?  Why would you choose only one?

Reviewer 4: I am not sure there is clarity here on what SAN is.

We have a client (the gateway) and server (the application). We want to use TLS to provide authentication, encryption, and integrity. SAN is what provides authentication. 

The client connects to the server and the server presents a certificate. The certificate has a SAN (usually - some only use CN but we mentioned it was deprecated 25 years ago so for all intents and purposes we can assume SAN is present to keep things simple). For a real world example, when you connect to google.com the SAN would be google.com.

Now the client wants to verify the server is who they expect (this is the authentication aspect of TLS; without this we only have encryption). First they verify the CA that signed the certificate is trusted (otherwise the data is not trusted). Then they verify who they are (the SAN) is aligned with their expectations.

In a browser scenario the expectation is always the DNS name matches the SAN.

In gateway, I don't think there is an obvious implicit expectation. We can default to the .cluster.local name perhaps but that is not going to be sufficient in all cases

Reviewer 5: I as an end-user would like to provide a list of host names here (SAN) for Upstream TLS verification

xref: https://github.com/projectcontour/contour/issues/5520

Author: @howardjohn I suppose if I just delete SAN as an implementation-dependent option, that would close this issue. Sound good?

Reviewer 6: Presumably, implementations are verifying _something_ today right? Since this GEP is going in as provisional, I think SAN verification can remain implementation-specific for now; however, I do agree that there needs to be a SAN list before this moves to implementable

Reviewer 4: SAN validation is a very very hard requirement. Without it we are basically doing `--insecure-skip-verify`

Author: > In a browser scenario the expectation is always the DNS name matches the SAN.
> 
> In gateway, I don't think there is an obvious implicit expectation. We can default to the .cluster.local name perhaps but that is not going to be sufficient in all cases

Here we're dealing only with the connection from gateway to backend.  I expect that one of the hostnames on the HTTPRoute matches the SAN.  Is that too much to expect?

Reviewer 4: The HTTPRoute hostname is the public internet facing hostname, not the internal service. So I'm a route from example.com->example-svc, I would Not expect the workload to have a cert for example.com

it could be example-svc.ns.svc.cluster.local maybe - but there is no standard

Author: Isn't that very service-mesh specific to not expect the pod to have a cert for example.com?  Alternatively, with SANs, it could list example.com and example-svc.ns.svc.cluster.local.

Reviewer 4: It is not related to service mesh in any way. I wouldn't expect them to have a cert for example.com since the Gateway itself has that. And it's not clear why you would serve the same cert in 2 places?

Reviewer 4: To be fair they could have any cert - even google.com (with their own ca, of course). So anything is *possible*

Reviewer: The internal workload on the pod can serve requests either from a Gateway or another internal workload (the mesh scenario) and in both cases is likely to serve the same certificate (unless SNI is used?) which is more likely to be example-svc and not example.com.

[RFC 8446 TLS 1.3](https://datatracker.ietf.org/doc/html/rfc8446#appendix-B.4).
- Version: specifies the minimum TLS version that the connection may use (default TLSv1.2)
- Visibility: the visibility level could be all or none, indicating that if the connection failed due to validation
failures, it would drop the connection silently if the visibility level were none, and report an error if the visibility
level were all.  (default to all)

```go
// TLSConnectionPolicy provides a way to publish TLS configuration
// that enables a gateway client to connect to a backend pod.
type TLSConnectionPolicy struct {
    metav1.TypeMeta   `json:",inline"`
    metav1.ObjectMeta `json:"metadata,omitempty"`

    // Spec defines the desired state of TLSConnectionPolicy.
    Spec TLSConnectionPolicySpec `json:"spec"`

    // Status defines the current state of TLSConnectionPolicy.
    Status TLSConnectionPolicyStatus `json:"status,omitempty"`
}

// TLSConnectionPolicySpec defines the desired state of
// TLSConnectionPolicy.
// Note: there is no Override or Default policy configuration.
type TLSConnectionPolicySpec struct {

>     // TargetRef identifies an API object to apply policy to.

Author: Services are the only valid API target references.

>     TargetRef gatewayv1a2.PolicyTargetReference `json:"targetRef"`

Reviewer 2: I'm really interested in if we can/should extend the TargetRef in the policy attachment GEP to optionally include port. It feels really strange to have port outside of the TargetRef because it's really just targeting a subset of that resource.

Reviewer 7: I like the term `section` that can be used to refer to a `listener` name within Gateway, `rule` name within xRoute, `port` name within Service

Author: I'd back extending TargetRef in the policy attachment GEP to optionally include port.

Reviewer 2: Maybe what we need to say is:

1. All policy resources must include TargetRef with the fields specified here: https://github.com/kubernetes-sigs/gateway-api/blob/a33a934af9ec6997b34fd9b00d2ecd13d143e48b/apis/v1alpha2/policy_types.go#L24-L41
2. Policy resources _may_ also choose to include `SectionName` and/or `Port` in TargetRef following the same mechanics of `ParentRef`

Then we can provide additional `TargetRef` types that extend the base `TargetRef` with the different possible combinations of these fields. That feels like it could be handled in a separate smaller update to the policy GEP.

Author: I added an issue for this: https://github.com/kubernetes-sigs/gateway-api/issues/2147

Author: I will add a small mention in the GEP, as well.

Reviewer 2: Thanks @candita! I think this can mostly be resolved since Port has been removed from the GEP and we have a follow up issue for this. Do you mind adding some kind of AI/follow up in this GEP that links to this issue and provides a bit of context about this policy targeting the entirety of a Service?

Author: Added to Open Questions section.

    // TLS contains TLS connection policy configuration.
    TLS *TLSConnectionPolicyConfig `json:”tls”`
}

// TLSConnectionPolicyConfig contains TLS connection policy configuration.
type TLSConnectionPolicyConfig struct {
    // TrustedCACertRefs contains one or more references to
    // Kubernetes objects that contain TLS certificates, which are

>     // used to establish a TLS handshake between the gateway and

Author: It depends, so it can be both.  It is one or more reference to a certificate.  When there is a chain of certificates, one or more can be CA/IntermediateCA.

    // backend pod.
    //
    // A single TrustedCACertRef to a Kubernetes Secret has "Core"
    // support.  Implementations MAY choose to support attaching
    // multiple certificates to a backend, but this behavior is 
    // implementation-specific.
    //
    // References to a resource in a different namespace are 
    // invalid.
    // 
    // This field is required for any TLSConnectionPolicyConfig.
    // 
    // Support: Core - A single reference to a Kubernetes Secret of type kubernetes.io/tls
    //
    // Support: Implementation-specific (More than one reference or other resource types)
    //
    // +kubebuilder:validation:MaxItems=64
    // +kubebuilder:validation:MinItems=1
    TrustedCACertRefs []SecretObjectReference `json:”trustedCACertRefs”`

>     // Port is the network port that the implementation watches to
>     // know if the connection should be TLS and the targetRef’s
>     // certificate should be validated by the certs in TrustedCACertRefs

Reviewer 2: As a minor nit, I think implementations will essentially apply this policy to anywhere they're mapping a BackendRef that matches the combination of the TLSConnectionPolicy TargetRef and Port. This is not great, but maybe something like this would be a bit clearer, very open to improvements/changes though:

```suggestion
    // Port is the network port of the target. When a target matches a BackendRef,
    // this Policy should apply, resulting in the certificate served by the backend
    // being validated by the certs in TrustedCACertRefs.
```

Author: Pending a decision on removing Port.

Reviewer 2: x-ref https://github.com/kubernetes-sigs/gateway-api/pull/2113#discussion_r1239035806

    // If empty, then all ports for the targetRef are watched to know
    // if the connection should be TLS, the targetRef’s certificate
    // should be validated by the certs in TrustedCACertRefs, and a
    // status delivered in the response for validation failures.

>     Port PortNumber `json:port,omitempty`

Reviewer: So the target of this policy is a `Service` and as far as I know the `Service` will have a `targetPort` (or defaulted to service port?). Can we just use that instead so this port is not needed? Otherwise how does the implementation work when the `targetPort` in the `Service` and the `Port` here are different?

Reviewer 2: In Gateway API, a port reference is traditionally part of a Service reference (BackendRef is best example) and it refers to the Port the Service is listening on, not the targetPort (port the Service backends are listening on). The reason this is needed is that it's fairly common for Kubernetes Services to listen on multiple ports and protocols.

So given the following example, this would be referring to port 80, and specifically backendRefs that are targeting this Service on port 80.

```
apiVersion: v1
kind: Service
metadata:
  name: my-service
spec:
  selector:
    app.kubernetes.io/name: MyApp
  ports:
    - protocol: TCP
      port: 80
      targetPort: 9376
```

Reviewer: > ...it's fairly common for Kubernetes Services to listen on multiple ports and protocols.

Got it. So in the case where a Kubernetes service listens on a single port, the `Port` field here (on line 243) has to match the service port. Put another way, the port can be inferred in such a case. Just confirming my understanding

    // conditions using the `TLSConnectionPolicyConditionType` and
    // `TLSConnectionPolicyConditionReason` constants so that 
    // operators and tools can converge on a common vocabulary to 
    // describe TLSConnectionPolicy state.
    // Known condition types are:
    // 
    // * “Accepted”
    // 
    // +optional
    // +listType=map
    // +listMapKey=type
    // +kubebuilder:validation:MaxItems=8
    // +kubebuilder:default={type: "Accepted", status: "Unknown", reason:"Pending", message:"Waiting for validation", lastTransitionTime: "1970-01-01T00:00:00Z"}
    Conditions []metav1.Condition `json:"conditions,omitempty"`
}

// TLSConnectionPolicyConditionType is the type of a condition used
// as a signal by TLSConnectionPolicy.  This type should be used with
// the TLSConnectionPolicyStatus.Conditions field.
type TLSConnectionPolicyConditionType string

//  TLSConnectionPolicyConditionReason is a reason that explains why a
// particular TLSConnectionPolicyConditionType was generated.
type TLSConnectionPolicyConditionReason string

const (
    // This condition indicates that the TLSConnectionPolicy has been
    // accepted as valid.

>     // Possible reason for this condition to be True is:
>     // * “Validated”
>     // Possible reasons for this condition to be False are:
>     // * “Invalid”
>     // * “Pending”

Reviewer 2: A couple suggested changes here - mostly list formatting for the sake of our reference docs, but also I think that we usually have the positive reason match the condition name.

```suggestion
    // Possible reason for this condition to be True is:
    //
    // * “Accepted”
    //
    // Possible reasons for this condition to be False are:
    //
    // * “Invalid”
    // * “Pending”
```

    TLSConnectionPolicyConditionAccepted TLSConnectionPolicyConditionType = “Accepted”

	// This reason is used with the “Accepted” condition when the condition is true.
    TLSConnectionPolicyReasonAccepted TLSConnectionPolicyConditionReason = “Valid”

    // This reason is used with the “Accepted” condition when the TLSConnectionPolicy is invalid, e.g. crossing namespace boundaries.
    TLSConnectionPolicyReasonInvalid TLSConnectionPolicyConditionReason = “Invalid”

    // This reason is used with the “Accepted” condition when the TLSConnectionPolicy is pending validation.
    TLSConnectionPolicyReasonPending TLSConnectionPolicyConditionReason = “Pending”
)
```

## How a client behaves

This table describes the effect that a TLSConnectionPolicy has on a Route.  There are only two cases where the
TLSConnectionPolicy will signal a Route to connect to a backend using TLS, an HTTPRoute with a backend that is targeted
by a TLSConnectionPolicy, either with or without listener TLS configured.  (There are a few other cases where it may be

> possible, but is purposely marked “not supported” due to a desire for less confusion on the assigned purpose of each of

Reviewer 4: What does an implementation do for "not supported"?

Author: In my understanding, these are invalid configurations, so they wouldn't pass validation.

Reviewer 4: how can they fail validation? The configuration and the route are separate objects

Author: That is true, I didn't think it through.  So for "not supported", the intention really is that this is not supported, at least not in this proposal, where we're only covering HTTPRoute.

Reviewer 2: This seems like something that would benefit from both nested conditions (as suggested in https://github.com/kubernetes-sigs/gateway-api/pull/2113/files#r1239031972) and the SupportedFeatures GEP from @LiorLieberman. We can say that every implementation that claims supports for this policy needs to populate status on the policy describing which Routes it is being implemented for (and _maybe_ also which Routes it is not being implemented for).

Author: How about, 

> Every implementation that claims supports for TLSConnectionPolicy should document for which Routes it is being implemented.

the protocol-affiliated types of Routes.)

| Route Type | Gateway Config             | Backend is targeted by a TLSConnectionPolicy? | Connect to backend  with TLS? |
|------------|----------------------------|-----------------------------------------------|-------------------------------|
| HTTPRoute  | Listener tls               | Yes                                           | **Yes**                       |
| HTTPRoute  | No listener tls            | Yes                                           | **Yes**                       |
| HTTPRoute  | Listener tls               | No                                            | No                            |
| HTTPRoute  | No listener tls            | No                                            | No                            |
| TLSRoute   | Listener Mode: Passthrough | Yes                                           | No                            |
| TLSRoute   | Listener Mode: Terminate   | Yes                                           | Not supported                 |
| TLSRoute   | Listener Mode: Passthrough | No                                            | No                            |

> | TLSRoute   | Listener Mode: Terminate   | No                                            | No                            |

Author: No, the mesh case is not covered in this proposal.

> | TLSRoute   | Listener Mode: Terminate   | No                                            | No                            |

Reviewer 4: I don't think we should be accepting new APIs into this repo that don't take mesh (now a core use case of the project) into account. I am not talking about handling "mesh transport", but we can't just have no answer for what  mesh client that is implementing the API should do when they API is there. Even if it's saying "should always ignore it"

> | TLSRoute   | Listener Mode: Terminate   | No                                            | No                            |

Author: I spelled it out in Longer Term Goals that service mesh use cases are a worthy goal, but it's a goal that may need a different GEP for proper attention.  I hope we don't get into a habit where we approve/merge a preliminary GEP with clearly stated goals and then request to change the goals later.

> | TLSRoute   | Listener Mode: Terminate   | No                                            | No                            |

Reviewer 6: Can we add a small clarification to longer term goals to say that implementations that support those 3 use-cases should ignore TLSConnectionPolicy?

| TCPRoute   | Listener TLS               | Yes                                           | Not supported                 |
| TCPRoute   | No listener TLS            | Yes                                           | Not supported                 |
| TCPRoute   | Listener TLS               | No                                            | No                            |
| TCPRoute   | No listener TLS            | No                                            | No                            |
| UDPRoute   | Listener TLS               | N/A                                           | No                            |
| UDPRoute   | No listener TLS            | N/A                                           | No                            |
| UDPRoute   | Listener TLS               | N/A                                           | No                            |
| UDPRoute   | No listener TLS            | N/A                                           | No                            |

> | GRPCRoute  | N/A                        | N/A                                           | No                            |

Reviewer 4: should probably be the same as HTTP?

Author: I'm not proposing any other changes in behavior except for HTTPRoute, so I'm not in a hurry to make GRPCRoute have the same behavior as HTTPRoute and be changed as well.

Reviewer 6: That feels fine for provisional but GRPCRoute should probably be supported for implementable. Not blocking right now though IMO

Author: I would request to wait until GRPCRoute itself is no longer experimental.   https://gateway-api.sigs.k8s.io/api-types/grpcroute/

 Is that acceptable?

Reviewer 2: I think this may actually be pretty straightforward and identical to HTTP as @howardjohn suggested. @gnossen can correct me here, but I think we should expect the same config from TLS from Gateway -> Backend to apply equally to HTTPRoute and GRPCRoute. 

As far as the experimental status of GRPCRoute, it's hard to get right. We've gotten into situations in upstream Kubernetes where we let in a bunch of incompatible alpha features and it became quite a pain to graduate any of them past alpha. In this case with GRPCRoute being fairly stable at this point, I'd rather cover it here unless it ends up adding significant complexity.

Author: I will mark it "implementation-dependent" intstead of "Yes" in the "Connect to backend with TLS?" column.

Reviewer 8: I agree with @robscott . Disregarding any issues of experimental state, I think the values for GRPCRoute should be the same as for HTTPRoute in this table.

## Request Flow

One additional step would be added to the typical client/gateway API request flow for a gateway implemented using a
reverse proxy. This is shown as step 6 below.

1. A client makes a request to http://foo.example.com.
2. DNS resolves the name to a Gateway address.
3. The reverse proxy receives the request on a Listener and uses the Host header to match an HTTPRoute.
4. Optionally, the reverse proxy can perform request header and/or path matching based on match rules of the HTTPRoute.
5. Optionally, the reverse proxy can modify the request, i.e. add/remove headers, based on filter rules of the HTTPRoute.

> 6. __(New) Optionally, the reverse proxy can determine the outcome of verifying the cert served by the backend, based on

Reviewer 4: nit: its a bit confusing to say we are verifying the cert served, but not mentioning that we are adding TLS in the first place

role, which violates the role-oriented design principle of Gateway API. As mentioned in Non-goal #7, providing a
mechanism for the cluster operator gateway role to override gateway to backend TLS settings is not covered here, but can
be addressed in a future update should the need arise.

TLSConnectionPolicy is defined as a Direct Policy Attachment without defaults or overrides, applied to a Service that
accesses the backend in question, where the TLSConnectionPolicy resides in the same namespace as the Service it is
applied to.  The TLSConnectionPolicy and the Service must reside in the same namespace in order to prevent the
complications involved with sharing trust across namespace boundaries.  By choosing the Service resource rather than the
Route resource, we can reuse the same TLSConnectionPolicy for all the different Routes that might point to this Service.
For the use case where certificates are stored in their own namespace, users may create Secrets and use ReferenceGrants
for a TLSConnectionPolicy-to-Secret binding.

In the API defined here, the definition of TrustedCACertRefs follows a convention established by TLSRoute in

> https://github.com/kubernetes-sigs/gateway-api/blob/main/apis/v1beta1/gateway_types.go#L340

Reviewer 9: Would it better to use a permalink here in case `gateway_types.go` changes in the future?

failures, it would drop the connection silently if the visibility level were none, and report an error if the visibility
level were all.  This is left as implementation-dependent.
- Server Name Indication: enables passing of the server name through server name indication, in the TLS transaction, to
assist with selection of certificates when several hosts share the same IP address. (default to enabled)
- Subject Alternative Name certificates: enable the use of a single certificate that can serve multiple domains.
(default to enabled)
- Version: specifies the minimum TLS version that the connection may use (default TLSv1.2)
- Ciphers: specifies enabled ciphers to be used in TLS exchanges. (default to align with TLS versions 1.1, 1.2, and/or
1.3 as described in [RFC4346 TLS 1.1](https://datatracker.ietf.org/doc/html/rfc4346#appendix-A.5),
[RFC 5246 TLS 1.2](https://datatracker.ietf.org/doc/html/rfc5246#appendix-A.5), and
[RFC 8446 TLS 1.3](https://datatracker.ietf.org/doc/html/rfc8446#appendix-B.4).
- Version: specifies the minimum TLS version that the connection may use (default TLSv1.2)

> - Visibility: the visibility level could be all or none, indicating that if the connection failed due to validation

Author: Fixed.

}

// TLSConnectionPolicySpec defines the desired state of
// TLSConnectionPolicy.
// Note: there is no Override or Default policy configuration.
type TLSConnectionPolicySpec struct {
    // TargetRef identifies an API object to apply policy to.
    TargetRef gatewayv1a2.PolicyTargetReference `json:"targetRef"`

    // TLS contains TLS connection policy configuration.
    TLS *TLSConnectionPolicyConfig `json:”tls”`
}

// TLSConnectionPolicyConfig contains TLS connection policy configuration.
type TLSConnectionPolicyConfig struct {
    // TrustedCACertRefs contains one or more references to
    // Kubernetes objects that contain TLS certificates, which are
    // used to establish a TLS handshake between the gateway and
    // backend pod.
    //
    // A single TrustedCACertRef to a Kubernetes Secret has "Core"
    // support.  Implementations MAY choose to support attaching
    // multiple certificates to a backend, but this behavior is 
    // implementation-specific.
    //
    // References to a resource in a different namespace are 
    // invalid.
    // 
    // This field is required for any TLSConnectionPolicyConfig.
    // 

>     // Support: Core - A single reference to a Kubernetes Secret of type kubernetes.io/tls

Reviewer 7: afaik a `kubernetes.io/tls` forces the secret to contain `tls.key` and `tls.crt`, but the `tls.key` is not required for TLS origination

Author: @arkodg do you have a suggestion or change for the `Support:`?

Reviewer 7: maybe we could try and relax this constraint, i.e. a `kubernetes.io/tls` is valid if a `tls.key` OR `tls.crt` key is set. 
cc @robscott

Reviewer 7: or the CA Cert can be inputted as a ConfigMap (local to the namespace), and in the case of mTLS, the client's key and cert can be inputted as a `kubernetes.io/tls` Secret

Reviewer: Elsewhere I commented that we should allow/support mTLS to be originated from the gateway so I like the idea of `kubernetes.io/tls` Secret

Reviewer 2: I've chatted a bit with @arkodg about this, and I think I like the idea of using something other than secret to story a CA Cert since that's not really a secret. One of the simplest paths forward would be to use a ConfigMap for CA Cert and a `kubernetes.io/tls` Secret for the client key and cert for mTLS. 

Alternatively, I'd chatted with @enj in the past about developing a different kind of resource under the scope of Gateway API that was meant for storing certs. This could potentially help with the issue we have today where most Ingress/Gateway controllers are deployed with read access to all secrets in the cluster by default. I think the ongoing ReferenceGrant work in sig-auth may make this idea irrelevant longer term, but thought it was at least mentioning in this context.

Reviewer: BTW, I just realized an mTLS client requires 3 things: a private key, identity cert (public key) and a trust store to validate peer (server) cert. If `kubernetes.io/tls` contains `tls.key` and `tls.crt` then we need one more thing (may be the Cluster trust bundle @howardjohn mentioned?) for full mTLS client side cert config.

Author: @robscott 
> I like the idea of using something other than secret to story a CA Cert since that's not really a secret. One of the simplest paths forward would be to use a ConfigMap for CA Cert and a kubernetes.io/tls Secret for the client key and cert for mTLS.

We could change this to ConfigMap for CA Certs now and add kubernetes.io/tls Secret to a different PolicyAttachment, like mTLSConnectionPolicy I mentioned earlier.

> Alternatively, I'd chatted with @enj in the past about developing a different kind of resource under the scope of Gateway API that was meant for storing certs. This could potentially help with the issue we have today where most Ingress/Gateway controllers are deployed with read access to all secrets in the cluster by default. I think the ongoing ReferenceGrant work in sig-auth may make this idea irrelevant longer term, but thought it was at least mentioning in this context.

It feels out of scope for this proposal, but I can open another discussion or issue to develop a different kind of resource for certs.

Reviewer 2: Yep, that makes sense to me - a well-defined field in a ConfigMap for now with a documented plan to move to a new resource type longer term.

Reviewer 6: +1 to ClusterTrustBundle in the long-term, but well-formatted configmap for now makes sense

Author: Made some changes for this.

> ## API

Reviewer 10: We talked about this one in the community sync today: we want to at least acknowledge mTLS use cases as a part of this doc, even if it is ultimately to say they wont be in scope (and why). We discussed making this an "unresolved" / "todo" so that it doesn't have to hold up this specific PR but it does need to hold us from moving out of `Provisional` state.

Author: Please see https://github.com/kubernetes-sigs/gateway-api/pull/2113/files/5d84379d2aa570958ce6dbd3fb8cf187447e9f95#r1231560568 and let me know if that is the discussion that needs to be a part of the proposal.

I have documented the mTLS use case are longer term goals that require more attention than this GEP could provide.  https://github.com/kubernetes-sigs/gateway-api/pull/2113/files/5d84379d2aa570958ce6dbd3fb8cf187447e9f95#diff-fbb3fdd16d81d10c31ffcac7b57c3107034f697fcf1c7b291f5d9e7ec51a4125L26

Author: In the next update I will also push a longer explanation in the "Longer Term Goals" section.

because this persona handles resources involved with application access and configuration, such as Routes and Services.
Choosing any other role would move the application-related responsibility from the application developer role to that
role, which violates the role-oriented design principle of Gateway API. As mentioned in Non-goal #7, providing a
mechanism for the cluster operator gateway role to override gateway to backend TLS settings is not covered here, but can
be addressed in a future update should the need arise.

TLSConnectionPolicy is defined as a Direct Policy Attachment without defaults or overrides, applied to a Service that
accesses the backend in question, where the TLSConnectionPolicy resides in the same namespace as the Service it is
applied to.  The TLSConnectionPolicy and the Service must reside in the same namespace in order to prevent the
complications involved with sharing trust across namespace boundaries.  By choosing the Service resource rather than the
Route resource, we can reuse the same TLSConnectionPolicy for all the different Routes that might point to this Service.

> For the use case where certificates are stored in their own namespace, users may create Secrets and use ReferenceGrants

Reviewer 6: It may be further down, but I feel like we should probably specify whether or not implementations SHOULD or MUST require and/or respect a ReferenceGrant for cross-namespace Secret sharing with respect to TLSConnection policy (i.e. is an implementation that doesn't require a ReferenceGrant for cross-namespace sharing non-conformant).

Author: @keithmattix I documented that references to a different namespace are invalid here: https://github.com/kubernetes-sigs/gateway-api/pull/2113/files/5d84379d2aa570958ce6dbd3fb8cf187447e9f95#diff-fbb3fdd16d81d10c31ffcac7b57c3107034f697fcf1c7b291f5d9e7ec51a4125R225-R226

Do you mean something else?  Or do I need to make TLSConnectionPolicy a namespaced resource?

Reviewer 6: Sorry for the delayed response; either one of your comments addresses my concern. I'll defer to others like @robscott on the partial requirement of ReferenceGrant

Reviewer 2: This is an interesting conversation and at least tangentially related to my other comments. Do we expect there to be a different CA cert per application? If so, I can't imagine that cross-namespace references would be useful for information like this that is not intended to be secret.

Or do we expect there to be a small set of CAs that are used for all certs served by backends in a cluster? If that's the case, maybe we shouldn't be configuring these per Service and instead have a broader target for this config.

Maybe there's a third option where whatever certs are served by backends are using a small set of CAs, but not all backends or ports are serving certs and so although we could configure the CAs at a higher level, we'd still need to separately configure which ports on which backends TLS certs should be verified?

In any case, I'm not really seeing the use case for cross-namespace references here, but interested in what others think.

Reviewer 2: Going through other comments, it looks like @sanjaypujare was actually asking similar questions in https://github.com/kubernetes-sigs/gateway-api/pull/2113/files#r1233341363 and I didn't fully understand the request until I ended up writing out something awfully similar myself above.

Author: @robscott is there an issue we still need to address here?

Reviewer 2: Nope, I think we're good now.

- Server Name Indication: enables passing of the server name through server name indication, in the TLS transaction, to
assist with selection of certificates when several hosts share the same IP address. (default to enabled)
- Subject Alternative Name certificates: enable the use of a single certificate that can serve multiple domains.
(default to enabled)
- Version: specifies the minimum TLS version that the connection may use (default TLSv1.2)
- Ciphers: specifies enabled ciphers to be used in TLS exchanges. (default to align with TLS versions 1.1, 1.2, and/or
1.3 as described in [RFC4346 TLS 1.1](https://datatracker.ietf.org/doc/html/rfc4346#appendix-A.5),
[RFC 5246 TLS 1.2](https://datatracker.ietf.org/doc/html/rfc5246#appendix-A.5), and
[RFC 8446 TLS 1.3](https://datatracker.ietf.org/doc/html/rfc8446#appendix-B.4).
- Visibility: the visibility level could be all or none, indicating that if the connection failed due to validation
failures, it would drop the connection silently if the visibility level were none, and report an error if the visibility
level were all.  (default to all)

```go
// TLSConnectionPolicy provides a way to publish TLS configuration

> // that enables a gateway client to connect to a backend pod.

Reviewer 9: Should we be clear in documenting the support level for `TLSConnectionPolicy`? Like how we do for the [ReferenceGrant](https://github.com/kubernetes-sigs/gateway-api/blob/cc80822bc68f6d1c27f9103c837f65cf28208907/apis/v1beta1/referencegrant_types.go#L43) struct. Is `TLSConnectionPolicy` expected to be supported universally?
```suggestion
// that enables a gateway client to connect to a backend pod.
// 
// Support: Core
```

Author: I can add it.  I have no opinion.

type TLSConnectionPolicy struct {
    metav1.TypeMeta   `json:",inline"`
    metav1.ObjectMeta `json:"metadata,omitempty"`

    // Spec defines the desired state of TLSConnectionPolicy.
    Spec TLSConnectionPolicySpec `json:"spec"`

    // Status defines the current state of TLSConnectionPolicy.
    Status TLSConnectionPolicyStatus `json:"status,omitempty"`
}

// TLSConnectionPolicySpec defines the desired state of
// TLSConnectionPolicy.
// Note: there is no Override or Default policy configuration.
type TLSConnectionPolicySpec struct {
    // TargetRef identifies an API object to apply policy to.
    // Services are the only valid API target references.

>     TargetRef gatewayv1a2.PolicyTargetReference `json:"targetRef"`

Reviewer: Can this be an array so we can create just one `TLSConnectionPolicy` to target multiple Services?

Author: I'm not convinced about this use case, can you provide more detail?  People set up Services that would share the same certificates between different applications/pods/endpoints?

Reviewer: You don't have to share the actual certificates (or the IDs in the certificates) between services. But the configuration in the `TLSConnectionPolicy` could be common or same among services. For example, the `SecretObjectReference` could point to a type of `Secret` that is just a pointer to a fixed file location and that's common for all services even if the actual file content at that location is different for different services. Does that make sense?

Reviewer 2: This is following the current policy attachment guidelines which only allow one target ref per instance of a policy resource. There's been some discussion about expanding that, but it's already quite complex. Until we can make it easier to reason about the existing system, we're trying to avoid adding any more complexity - x-ref: https://github.com/kubernetes-sigs/gateway-api/issues/2107#issuecomment-1587856919.

    // TLS contains TLS connection policy configuration.

>     TLS *TLSConnectionPolicyConfig `json:”tls”`

Reviewer 5: This might be better as a list that's keyed off the PortNumber

Then it's possible to have different TLS configurations for different ports in a single Service. Unless you were specifically looking to have those be distinct policies?

Author: The guidelines are only one target ref per policy attachment.  Maybe this also depends on whether we create a new type of target ref that includes port or ports.
x-ref https://github.com/kubernetes-sigs/gateway-api/pull/2113/files/5d84379d2aa570958ce6dbd3fb8cf187447e9f95#r1231594914

Reviewer 3: To close this one out, we're adding `sectionName` to Policy TargetRef - which will allow you to target the port's `name` field.

}

// TLSConnectionPolicyConfig contains TLS connection policy configuration.
type TLSConnectionPolicyConfig struct {
    // TrustedCACertRefs contains one or more references to
    // Kubernetes objects that contain TLS certificates, which are
    // used to establish a TLS handshake between the gateway and
    // backend pod.
    //
    // A single TrustedCACertRef to a Kubernetes Secret has "Core"
    // support.  Implementations MAY choose to support attaching
    // multiple certificates to a backend, but this behavior is 
    // implementation-specific.
    //
    // References to a resource in a different namespace are 
    // invalid.
    // 
    // This field is required for any TLSConnectionPolicyConfig.
    // 
    // Support: Core - A single reference to a Kubernetes Secret of type kubernetes.io/tls
    //

>     // Support: Implementation-specific (More than one reference or other resource types)

Reviewer 4: I realize this could be done through implementation specific things, but is it possible to have an option to use the system trusted certificates? This is how most TLS configurations i have seen work - I don't have to tell chrome, curl, etc to use `/etc/ssl/certs/ca-certificates.crt` for example.

This is important since these do not exist as Secrets. Rather, they are... somewhere... Every client is going to have some, but they may be in different places (there are like 8 different well-known filepaths in https://golang.org/src/crypto/x509/root_linux.go for example, so systems do not use files. etc).

One option is to make this the default if the list is unset.

    //
    // +kubebuilder:validation:MaxItems=64
    // +kubebuilder:validation:MinItems=1
    TrustedCACertRefs []SecretObjectReference `json:”trustedCACertRefs”`

    // Port is the network port that the implementation watches to
    // know if the connection should be TLS and the targetRef’s
    // certificate should be validated by the certs in TrustedCACertRefs
    // If empty, then all ports for the targetRef are watched to know
    // if the connection should be TLS, the targetRef’s certificate
    // should be validated by the certs in TrustedCACertRefs, and a
    // status delivered in the response for validation failures.

>     Port PortNumber `json:port,omitempty`

Reviewer 5: Do we care about letting policy authors use port names and not just numbers?

Similar to how [K8s Service ](https://kubernetes.io/docs/reference/kubernetes-api/service-resources/service-v1/)`ports.targetPort` is `IntOrString` ?

Author: Let's decide on whether we need a new object that includes Port, then we can decide on this.

}

// TLSConnectionPolicyStatus defines the observed state of TLSConnectionPolicy.
type TLSConnectionPolicyStatus struct {
    // Conditions describe the current conditions of the TLSConnectionPolicy.
    //
    // Implementations should prefer to express TLSConnectionPolicy
    // conditions using the `TLSConnectionPolicyConditionType` and
    // `TLSConnectionPolicyConditionReason` constants so that 
    // operators and tools can converge on a common vocabulary to 
    // describe TLSConnectionPolicy state.
    // Known condition types are:
    // 
    // * “Accepted”
    // 
    // +optional
    // +listType=map
    // +listMapKey=type
    // +kubebuilder:validation:MaxItems=8
    // +kubebuilder:default={type: "Accepted", status: "Unknown", reason:"Pending", message:"Waiting for validation", lastTransitionTime: "1970-01-01T00:00:00Z"}

>     Conditions []metav1.Condition `json:"conditions,omitempty"`

Reviewer 11: Assuming multiple implementations running in the cluster,  will they start conflicting with each other when updating the status of the resource?

Author: It's an interesting question.  I haven't seen the same question applied to other proposals, is it pertinent here because TLSConnectionPolicy isn't namespaced?  I wouldn't expect multiple implementations to interfere with each other, @robscott @shaneutt @youngnick are you aware if there's a GEP addressing multiple-implementations co-existing on a cluster?  I'd expect that each implementation knows how to access only those objects that it created, but I don't know for sure.

Reviewer 2: Yes, that's a very good point. We probably need to introduce some form of nesting here, similar to [Route status](https://github.com/kubernetes-sigs/gateway-api/blob/a33a934af9ec6997b34fd9b00d2ecd13d143e48b/apis/v1beta1/shared_types.go#L361) where each set of conditions is per-Parent. I'm not sure what we'd call that, maybe `TargetParents` instead of just `Parents`? 

Open to ideas, but agree that it's a gap in our policy attachment model. Right now it doesn't really define how status should be handled for policies that are implemented by multiple controllers.

Reviewer 6: In general, assuming multiple controllers feels like a best practice for Gateway API resources _unless_ the resource in question has some field that effectively limits it to a single controller (e.g. `gatewayClassName` on Gateway). I'm in favor of making the change in this GEP specifically as well as a general change to the policy attachment model

Author: Added to Open Questions section.

}

// TLSConnectionPolicyConditionType is the type of a condition used
// as a signal by TLSConnectionPolicy.  This type should be used with
// the TLSConnectionPolicyStatus.Conditions field.
type TLSConnectionPolicyConditionType string

//  TLSConnectionPolicyConditionReason is a reason that explains why a
// particular TLSConnectionPolicyConditionType was generated.
type TLSConnectionPolicyConditionReason string

const (
    // This condition indicates that the TLSConnectionPolicy has been
    // accepted as valid.
    // Possible reason for this condition to be True is:
    //
    // * “Accepted” 
    // 

>     // Possible reasons for this condition to be False are:

Reviewer 11: can `ResolvedRefs` condition be also used here? since the policy addresses external references - service, secrets

Author: Sounds reasonable, unless there are privacy issues.  As an implementor, what kind of condition reasons would you think should be there?  Should there be a general validation on the certificate so it could have an invalid cert reason?  Or does this store results of an analysis of reference grants between the policy and targets?

Reviewer 2: I think _most_ implementations would only look at this if they were implementing a Service that was targeted by one of these policies, so I don't think `ResolvedRefs` would be relevant in that case, but I agree that it likely would be very useful for the secret reference(s) that will be included.

Author: This one I will pass on unless there is further interest.

## API

To allow the gateway client to know how to connect to the backend pod, when the backend pod has its own
certificate, we implement a metaresource that is already mentioned as an example in
[GEP-713: Metaresources and PolicyAttachment](https://gateway-api.sigs.k8s.io/geps/gep-713/), and as a hypothetical in the
[Policy Attachment](https://gateway-api.sigs.k8s.io/references/policy-attachment/#direct-policy-attachment)
documentation.

This metaresource is named TLSConnectionPolicy.  In this document, because naming is hard, we chose to retain the
name TLSConnectionPolicy to advertise alignment with a previously discussed naming choice, but a new name may be
substituted without blocking acceptance of the content of the API change.

The selection of the applicable Gateway API persona is important in the design of this proposal, because each

> persona is assigned a role which handles specific Gateway API resources.  TLSConnectionPolicy is used by the application
> developer Gateway API persona to convey client certificate settings used in the TLS handshake from Gateway to backend,
> because this persona handles resources involved with application access and configuration, such as Routes and Services.

Reviewer 3: I think the important part here is to consider whose _intent_ we are representing in this API. I suggest something like this:
```suggestion
persona is assigned a role which handles specific Gateway API resources.  TLSConnectionPolicy is used by the application
developer Gateway API persona to signal what the application developer _expects_ connections to the application to look like from a TLS perspective. 
Only the application developer can know what the application expects, so this it's important that this configuration be owned by that persona.
```
The important part here is that the _application running inside the Pods_ expects connections to look a certain way - the TLSConnectionPolicy object is about having a way for the application developer to _explicitly_ describe that expectation.

We can use this way of looking at things to make calls about what should be included in the TLSConnectionPolicy.

* The _backend_ should already be managing its own cert somehow, so we shouldn't need to represent that here.
* The Gateway, as a TLS client, needs to know some things about how to connect to the service (most of these are handled automatically for _end user clients_ by the browser):
  * that TLS should be attempted at all (a browser will use the scheme to determine this, but Gateways don't have that information available _at an API level_ - it could be intuited from some config, but it should be explicitly configured.)
  * What CA should be used (a browser usually will use the system bundled CAs, and allow adding extras, it's _much_ more common to use either individual self-signed certificates or an internal CA for Gateway-to-internal-backend use cases)
  * What version of TLS should be expected (this is not strictly necessary, because TLS is a negotiated protocol, but it is _very_ nice to have because it will allow Gateway implementations to do things like disallow TLS 1.1, force only TLS 1.3, and so on).

While the Gateway as a TLS client does need to know if it should pass a client certificate, that should be configured on the Gateway, by the Gateway owner, so that the client and server have their chain of custody properly managed. (see https://github.com/kubernetes-sigs/gateway-api/pull/2113/files#r1268891063 for some more thoughts on this from me).

> because this persona handles resources involved with application access and configuration, such as Routes and Services.

Author: @robscott I have added to the API to describe an idea that if the backend requires TLS and the Gateway hasn't been configured for it, then it should be possible to use system trusted certs.  I think this addresses your concern.  I don't feel it is a problem for 1 and 2 to be done by the same person.

@youngnick I agree with what you've written here and can update the API.

> because this persona handles resources involved with application access and configuration, such as Routes and Services.

Author: The first two are already there:
That TLS should be attempted at all -> the TLSConfigPolicy is there for the targetRef.
What CA to use -> TrustedCACertRefs

I will add the TLS versions.

Choosing any other role would move the application-related responsibility from the application developer role to that
role, which violates the role-oriented design principle of Gateway API. As mentioned in Non-goal #7, providing a
mechanism for the cluster operator gateway role to override gateway to backend TLS settings is not covered here, but can

> be addressed in a future update should the need arise.

Author: I'm working on an update for this that ties in with updates for https://github.com/kubernetes-sigs/gateway-api/pull/2113#discussion_r1268822227 and https://github.com/kubernetes-sigs/gateway-api/pull/2113#discussion_r1268903825

> be addressed in a future update should the need arise.

Author: I'll be adding something like this to explain a future approach.

One idea is to use two types: ApplicationTLSConnectionPolicy,
and GatewayTLSConnectionPolicy, where the application developer is responsible for the former, the cluster operator is responsible for the latter, and the cluster operator may configure whether certain settings may be overridden by application developers.

- Subject Alternative Name certificates: allow authentication with a single certificate that can serve multiple domain
names that are intended to represent the same endpoint.  For example, these five SAN hosts: example.com, 
www.example.com, example.net, IPv4 x.x.x.x, and IPv6 x.x.x.x.x.x.x.x all resolve to the same endpoint.
(default to enabled)
- Version: specifies the minimum TLS version that the connection may use (default TLSv1.2)
- Ciphers: specifies enabled ciphers to be used in TLS exchanges. (default to align with default TLS version chosen for
Version, e.g. ciphers mentioned for TLS version 1.1, 1.2, or 1.3 as described in 
[RFC4346 TLS 1.1](https://datatracker.ietf.org/doc/html/rfc4346#appendix-A.5),
[RFC 5246 TLS 1.2](https://datatracker.ietf.org/doc/html/rfc5246#appendix-A.5), or
[RFC 8446 TLS 1.3](https://datatracker.ietf.org/doc/html/rfc8446#appendix-B.4).
- Visibility: the visibility level could be `all` or `none`, and represents that if the connection failed, 
the implmentation would drop the connection silently if the visibility level were `none`, or report an error if the

> visibility level were `all`.  (default to `all`)

Reviewer 2: I don't think this content is needed in the GEP, and they seem like they could be more controversial than they're worth. I'd rather leave them out until we actually provide a way to configure any of these settings.

Author: I will be keeping and expanding the discussion on AllowedSANs and TLSVersions, which have been added.

Thus, the following additions would be made to the Gateway API:

```go
// TLSConnectionPolicy provides a way to publish TLS configuration
// that enables a gateway client to connect to a backend pod.

> type TLSConnectionPolicy struct {

Reviewer 9: https://github.com/kubernetes-sigs/gateway-api/pull/2159#discussion_r1259991693 - seems like everyone was voting to not use `Policy` in the name. Posting here for visibility.

@robscott @howardjohn

Reviewer: Good callout - I had the same point. Is the `...Policy` here okay?

Reviewer 3: Both of these are metaresources that use the Policy Attachment pattern,, so I think that having `...Policy` should actually be a requirement. I've commented on the other issue to the same effect.

Reviewer 3: If anyone can come up with a better name, I welcome feedback, but I think that having something that clearly signals "this object is a metaresource that uses Policy Attachment" in the name is a must-have.

Reviewer 12: Since virtually all configs in K8S are also Policies - how does it help if all have this suffix ? Policy is not specific to Gateways, there are authz policies, etc. 

I don't mind if it is decided that all Gateway API resources have a suffix and is used in all, for discoverability. But maybe not a broad word as 'policy', and it must be clearly defined in the spec what is the word and that all resources must use it.

Author: In the interest of time, I kindly request that you open a new issue to discuss naming policies.

    metav1.TypeMeta   `json:",inline"`
    metav1.ObjectMeta `json:"metadata,omitempty"`

    // Spec defines the desired state of TLSConnectionPolicy.
    Spec TLSConnectionPolicySpec `json:"spec"`

    // Status defines the current state of TLSConnectionPolicy.
    Status TLSConnectionPolicyStatus `json:"status,omitempty"`
}

// TLSConnectionPolicySpec defines the desired state of
// TLSConnectionPolicy.
// Note: there is no Override or Default policy configuration.
//
// Support: Core
type TLSConnectionPolicySpec struct {
    // TargetRef identifies an API object to apply policy to.

>     // Services are the only valid API target references.

Reviewer 2: Thanks for removing port from the `TLSConnectionPolicyConfig` section! Do you mind clarifying here that this config applies to an entire resource and maybe add a note that in the future a SectionName or Port may be added to this TargetRef to allow for more granular application of this policy?

    TargetRef gatewayv1a2.PolicyTargetReference `json:"targetRef"`

    // TLS contains TLS connection policy configuration.
    TLS *TLSConnectionPolicyConfig `json:”tls”`
}

// TLSConnectionPolicyConfig contains TLS connection policy configuration.
type TLSConnectionPolicyConfig struct {
    // TrustedCACertRefs contains one or more references to
    // Kubernetes objects that contain PEM-encoded TLS certificates,
    // which are used to establish a TLS handshake between the gateway
    // and backend pod.
    //
    // If the TrustedCACertRefs is empty, then the system trusted
    // certificates should be used. If there are none, or the
    // implementation doesn't define system trusted certificates,

>     // then a TLS connection must fail.

Reviewer 2: I'm not really sure where to capture this, but I think we need to clearly state that the client (Gateway) is responsible for verifying the SAN is in the certificate served by the application.

    //
    // A single TrustedCACertRef to a Kubernetes ConfigMap has "Core"
    // support.  Implementations MAY choose to support attaching
    // multiple certificates to a backend, but this behavior is
    // implementation-specific.
    //
    // References to a resource in a different namespace are 
    // invalid.
    // 
    // Support: Core - An optional single reference to a Kubernetes
    // ConfigMapName.
    //
    // Support: Implementation-specific (No reference, more than one
    // reference, or other resource types.)
    //
    // +kubebuilder:validation:MaxItems=64
    // +optional

>     TrustedCACertRefs []ConfigMapName `json:”trustedCACertRefs”`

Reviewer 2: I think it's highly likely that in the future we'll support a different kind of resource to store certificates. To ensure we leave room for that, I'd rather use a type of reference that includes group and kind. We can/should still leave in the guidance that only ConfigMap has "Core" support here though.

Reviewer 2: We may actually want to consider making [ClusterTrustBundle](https://github.com/kubernetes/enhancements/tree/master/keps/sig-auth/3257-trust-anchor-sets#clustertrustbundle-object) "Core" support in the near future whenever it graduates to beta or beyond. It would be good to reference that somewhere in this GEP as a future addition and maybe even state in this guidance that we anticipate transitioning to ClusterTrustBundle as it becomes more widely available.

Author: @robscott I need more context for the type of reference that includes group and kind, and still leave in support for ConfigMap.

Reviewer 3: There's a a few ObjectReference types in `object_reference_types.go` in `v1beta1` - particularly LocalObjectReference may be useful there if we want to disallow cross-namespace CACertRefs. If we do want to allow cross-namespace references, probably the core k8s `ObjectReference` is the way to go (see https://pkg.go.dev/k8s.io/api/core/v1#ObjectReference)

In January 2023 we closed GEP-1282 and began a new discussion on enumerating TLS use cases in
[Gateway API TLS Use Cases](#references), for the purposes of a clear definition and separation of concerns.
This GEP is the outcome of the TLS use case #4 in
[Gateway API TLS Use Cases](#references) as mentioned in the Immediate Goals section above.

## API

To allow the gateway client to know how to connect to the backend pod, when the backend pod has its own
certificate, we implement a metaresource named `TLSConnectionPolicy`, that is previously mentioned as an example in
[GEP-713: Metaresources and PolicyAttachment](https://gateway-api.sigs.k8s.io/geps/gep-713/), and as a hypothetical in the

> [Policy Attachment](https://gateway-api.sigs.k8s.io/references/policy-attachment/#direct-policy-attachment)

Reviewer 3: This reference is gone since I updated that page to just point back to GEP-713 (so we only have one place to update).

Author: Updated.

that there was a failure in the certificate validation, this is left up to the implementation to return a response error
that is appropriate, such as one of the HTTP error codes: 400 (Bad Request), 401 (Unauthorized), 403 (Forbidden), or
other signal that makes the failure sufficiently clear to the requester without revealing too much about the transaction,
based on established security requirements.

All policy resources must include `TargetRef` with the fields specified
[here](https://github.com/kubernetes-sigs/gateway-api/blob/a33a934af9ec6997b34fd9b00d2ecd13d143e48b/apis/v1alpha2/policy_types.go#L24-L41).
In an upcoming [extension](https://github.com/kubernetes-sigs/gateway-api/issues/2147) to TargetRef, policy resources
_may_ also choose to include `SectionName` and/or `Port` in the TargetRef following the same mechanics as `ParentRef`.

TLSConnectionPolicySpec contains the `TargetRef` and `TLS` fields.  The `TLS` field is a `TLSConnectionPolicyConfig` and
contains `CertRefs`, `CACertRefs`, and  `Hostname`.
The names of the fields were chosen to facilitate discussion, but may be substituted without blocking acceptance of the
content of the API change.

> The `TrustedCA` contains the `CertRefs` and `CACertRefs` fields. CertRefs is a slice of

Reviewer 3: I know, naming is hard, but I think it's important that the field currently called `CACertRefs` does _not_ end with `Refs` - that is reserved in the API conventions for structs that contain a list of references to objects.

How about instead of `CACertRefs`, we have `UseExtraCerts` - I suggest this because it has two possible values (YAML shown for ease of typing):
- `trustedCAs.useExtraCerts: ""` (or unspecified) - don't use any extra certificates for Trusted CAs.
- `trustedCAs.useExtraCerts: "system"` - use the system certificates as extra trusted CAs.

This name _implies_ that it's possible to use `CertRefs` _with_ `UseExtraCerts` - which I think has undefined behavior at this time.

Other names I can think of:
- `AdditionalCerts` (YAML `trustedCAs.additionalCerts`) Also implies a merging behavior between `CertRefs` and `AdditionalCerts`.
- `AlternativeCerts` (YAML `trustedCAs.alternativeCerts`) Implies that you can only choose _either_ `CertRefs` _or_ `AdditionalCerts` - they are mutually exclusive.

Whatever name we choose, we need to ensure that the behavior when you set both is defined. ("That's a syntax error" is okay, as is "the lists must be merged" or other possible answers, we just need to pick one, which we can also mark as temporary - given that this is `Provisional`.)

Again, naming is hard. ping @robscott as well as @candita.

Reviewer 2: Yeah I've struggled with naming on this one as well. Of course I'd originally proposed `wellKnownRoots` but obviously that has it's flaws, and not all systems are necessarily going to be using "well-known" values. So _maybe_ we should just narrow down the scope of this field a bit. It seems _very unlikely_ that Kubernetes or Gateway API will ever try to maintain a common set of certificates that should be used for this purpose, so we can likely say that whatever values are supported here other than none/empty, are going to be at least somewhat implementation-specific. 

I think we should use the following guidelines when choosing a name here:

1. Avoid a name that locks us into whether/how this combines with `certRefs`. My personal opinion is that we should start with validation that ensures that only `certRefs` or this field can be set, and we can loosen that later if needed. I think this would rule out `additional`, `alternative`, or `extra`.
2. Avoid a boolean value so we're not conflicting with the [relevant Kubernetes API Conventions](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api-conventions.md#primitive-types).
3. Building on 2, allow for some theoretical room for expansion. I think the most likely value here would be domain-prefixed values if for some reasons implementations want to support predefined sets of CAs. Even this seems highly unlikely at this point, but want to at least leave room for it so we're not boxed in years from now.

With all that said, I'm struggling to actually find a good name that meets this criteria. Some more ideas:

A. `trustedCAs.systemCerts: "System"`
B. `trustedCAs.predefinedCerts: "System"`
C. `trustedCAs.standardCerts: "System"`

As far as whether the empty value should be explicit ("None") or just empty (nil pointer or ""), I tend to lean towards a default of "None" so there's absolutely no ambiguity here. If we'd rather avoid "None", I think I'd lean towards a nil pointer instead of "", but open to what others think.

Author: @youngnick  I changed the name of `CACertRefs`, didn't realize the Refs suffix had a semantic association.  Should I also change the name of `CertRefs`, which a slice of ConfigMapNames, or is that fitting?

@robscott and I had this conversation last night and I originally proposed `alternativeCerts` (amongst others) but I get Rob's point about preferring to use a name that doesn't lock us into how this field _combines_ with `certRefs`.  I chose `standardCerts`.  `certRefs` and `standardCerts` are intended to be mutually exclusive in the first round.

Regarding using a nil pointer instead of "" for `standardCerts`, I'd rather leave it "" so that the field is visible when yaml is examined.  If we don't normally do that for our users, I can make it a pointer.

Reviewer 3: Oh, I missed that `CertRefs` is a slice of strings that are ConfigMap names. Yes, that should be a `[]ObjectReference` instead, with a default of Configmap as the Kind (which puts "", meaning "core" as the Group). That way, you can specify a list of certificates in Configmaps like this:
```yaml
    ...
    certRefs:
    - name: CAcert1
    - name: CAcert2
    ...
```
or, you can also specify other objects like Secrets if you want. Anything _not_ a ConfigMap should be Implementation Specific support in the initial version though. This lets us change that _much_ more easily lately.  (Practically, to be able to set the defaults, you'll need to create another type like `SecretObjectReference` - probably `ConfigmapObjectReference`).

Using a nil pointer or "" as the default value makes no difference - because the field is optional and will have `omitempty` set, then a nil pointer (for a `*string` type) or an empty string (for a `string` type) will both be not necessary in input YAMLs, and hidden in output YAMLs. The only way you'll see the `standardCerts` field (which I agree is a good name) is if it's set to `system`. However, if you _do_ set `standardCerts: system`, then the `CertRefs` field also must be empty in this version, so it will also be hidden on output.

So you'll only ever see:
```yaml
    ...
    certRefs:
    - name: CAcert1
    ...
```
or
```yaml
    ...
	standardCerts: "system"
    ...
```
Which I think is what we want.

So, my requested changes here are:
- `CertRefs` should be a `[]ConfigmapObjectReference`, where that is an ObjectReference struct that defaults to Configmap.
- `standardCerts` should be a plain string field (not a `*string`), with a default value of "" that's optional.

Author: @youngnick we discussed offline, and I will make the requested change to `CertRefs` and do `StandardCerts` like this:

* standardCerts should be a `StandardCertType` pointer field with Enum validation, and one valid value: System.  An informal review showed that alias strings usually start with a capital letter, particularly condition types and reasons, so I used System rather than system.

// TLSConnectionPolicy.
// Note: there is no Override or Default policy configuration.
//
// Support: Core
type TLSConnectionPolicySpec struct {
    // TargetRef identifies an API object to apply policy to.
    // Services are the only valid API target references.
    // Note that this config applies to the entire referenced resource
    // by default, but this default may change in the future to provide
    // a more granular application of the policy.
    TargetRef gatewayv1a2.PolicyTargetReference `json:"targetRef"`

    // TLS contains TLS connection policy configuration.
    TLS *TLSConnectionPolicyConfig `json:”tls”`
}

// TLSConnectionPolicyConfig contains TLS connection policy configuration.
type TLSConnectionPolicyConfig struct {
    // CertRefs contains one or more references to
    // Kubernetes objects that contain PEM-encoded TLS certificates,
    // which are used to establish a TLS handshake between the gateway
    // and backend pod.
    //
    // If the CertRefs is empty, then the system trusted
    // certificates should be used. If there are none, or the
    // implementation doesn't define system trusted certificates,

>     // then a TLS connection must fail.

Reviewer 3: We should remove this because this behavior is now handled by the other field (`CACertRefs` here, but as I said above, I think it needs a new name).

Author: I think a user could accidentally try to use System certs but the system doesn't have them, so I think it should stay in.

If StandardCerts is set to "System" and there are no system trusted certificates or the implementation doesn't define system
trusted certificates, then the associated TLS connection must fail.

The `Hostname` field is required and is to be used to configure the SNI the Gateway should use to connect to the backend.
Implementations must validate that at least one name in the certificate served by the backend matches this field.
We originally proposed using a list of allowed Subject Alternative Names, but determined that this was [not needed in
the first round](https://github.com/kubernetes-sigs/gateway-api/pull/2113#issuecomment-1696127092),
but may be added in the future.

We originally proposed allowing the configuration of expected TLS versions, but determined that this was [not needed in
the first round](https://github.com/kubernetes-sigs/gateway-api/pull/2113#issuecomment-1696127092).

Thus, the following additions would be made to the Gateway API:

```go
import "sigs.k8s.io/gateway-api/apis/v1beta1"

// TLSConnectionPolicy provides a way to publish TLS configuration
// that enables a gateway client to connect to a backend pod.

> type TLSConnectionPolicy struct {

Author: Done.

    // CertRefs contains one or more references to
    // Kubernetes objects that contain PEM-encoded TLS certificates,
    // which are used to establish a TLS handshake between the gateway
    // and backend pod.
    //
    // If CertRefs is empty or unspecified, then StandardCerts must
    // be specified.  Only one of CertRefs or StandardCerts may be
    // specified, not both.
    //
    // If CertRefs is empty or unspecified, then system trusted
    // certificates should be used. If there are none, or the
    // implementation doesn't define system trusted certificates,
    // then a TLS connection must fail.
    //
    // References to a resource in a different namespace are
    // invalid.
    //
    // A single CertRef to a Kubernetes ConfigMap has "Core"
    // support.  Implementations MAY choose to support attaching
    // multiple certificates to a backend, but this behavior is
    // implementation-specific.
    // 
    // Support: Core - An optional single reference to a Kubernetes
    // ConfigMapName.
    //
    // Support: Implementation-specific (No reference, more than one
    // reference, or other resource types; service mesh may ignore.)
    //

>     // +kubebuilder:validation:MaxItems=64

Author: Done.

    // then a TLS connection must fail.
    //
    // Support: Core - An optional value to specify whether to use
    // system certificates or not.
    //
    // Support: Implementation-specific (In the absence of support
    // for usable system certs, may be ignored. Service mesh may ignore.)
    //
    // +optional
    // kubebuilder:default=""
    StandardCerts StandardCertType `json:"standardCerts"`

    // Hostname is the Server Name Indication that the Gateway uses to
    // connect to the backend.  It represents the fully qualified domain
    // name of a network host, as defined by RFC1123 - except that numeric
    // IP addresses are not allowed. Each label of the FQDN must consist
    // of lower case alphanumeric characters or '-', and must start and
    // end with an alphanumeric character.  No other punctuation is allowed.
    // Wildcard domain names are specifically disallowed.
    //
    // It specifies the hostname that may authenticate, and must be in the
    // certificate served by the matching backend.
    //
    // Support: Core - A required value used by the Gateway to connect to
    // the backend when a TLSConnectionPolicy is specified.

>     Hostname v1beta1.PreciseHostname `json:"hostname,omitempty"`

Author: Done.

for these use cases of the Gateway as a TLS Client:

- An explicit signal that TLS should be used by this connection.
- Which Certificate Authority (CA) to use, whether the system bundled CA, an internal CA, or self-signed.
- TLS version to expect.  This allows implementations to introduce features to disallow or force certain TLS versions.

TLSConnectionPolicy is defined as a Direct Policy Attachment without defaults or overrides, applied to a Service that
accesses the backend in question, where the TLSConnectionPolicy resides in the same namespace as the Service it is
applied to.  The TLSConnectionPolicy and the Service must reside in the same namespace in order to prevent the
complications involved with sharing trust across namespace boundaries.  By choosing the Service resource rather than the
Route resource, we can reuse the same TLSConnectionPolicy for all the different Routes that might point to this Service.
For the use case where certificates are stored in their own namespace, users may create Secrets and use ReferenceGrants
for a TLSConnectionPolicy-to-Secret binding.  Implementations must respect a ReferenceGrant for cross-namespace Secret

> sharing to TLCConnectionPolicy, even if they don't for other cross-namespace sharing.

Reviewer 6: ```suggestion
sharing to TLSConnectionPolicy, even if they don't for other cross-namespace sharing.
```

    // implementation doesn't define system trusted certificates,
    // then a TLS connection must fail.
    //
    // A single TrustedCACertRef to a Kubernetes ConfigMap has "Core"
    // support.  Implementations MAY choose to support attaching
    // multiple certificates to a backend, but this behavior is
    // implementation-specific.
    //
    // References to a resource in a different namespace are 
    // invalid.
    // 
    // Support: Core - An optional single reference to a Kubernetes
    // ConfigMapName.
    //
    // Support: Implementation-specific (No reference, more than one
    // reference, or other resource types.)
    //
    // +kubebuilder:validation:MaxItems=64
    // +optional
    TrustedCACertRefs []ConfigMapName `json:”trustedCACertRefs,omitempty”`

    // AllowedSANs are the Subject Alternative Names that are allowed for
    // this TLS connection.  It represents the hostnames that may
    // authenticate, and must be in the received certificate.
    //
    // If AllowedSANs is set, then an implementation must validate that
    // the certificate provided by the pod represents a SAN in its list.
    //

>     // If AllowedSANs is empty, then any SAN is allowed.

Author: @howardjohn @keithmattix - I'm adding the `backendCABundle` that Nick mentioned in https://github.com/kubernetes-sigs/gateway-api/pull/2113#issuecomment-1651028525.   Could this be used if the
AllowedSANs field is absent, and the use case was self-signed certs?

>     // If AllowedSANs is empty, then any SAN is allowed.

Reviewer 4: In TLS we ought to either verify the SAN _or_ make a user explicitly opt into "Insecure Mode".

So I disagree with Nick's suggestion. `backendCABundle` is insecure-by-design and cannot be accepted IMO. TLS _must_ verify certificate fully: that means the CA but also properties of the cert, most notably the SANs.

>     // If AllowedSANs is empty, then any SAN is allowed.

Reviewer 12: Or follow the standards - SNI, SAN and authority set to the hostname of the backend - and verify standard DNS cert. 

SAN is required for spiffee or custom schemes. I don't think the standard DNS case needs anything special.

>     // If AllowedSANs is empty, then any SAN is allowed.

Reviewer 6: I think the ca bundle and SAN verification are orthogonal pieces of the same process. SAN verification says "even if we have the same trust chain, I need to know that I'm talking to the identity I'm intending to"

>     // If AllowedSANs is empty, then any SAN is allowed.

Author: @costinm just to be clear, are you talking about standard Server Name Indication (https://datatracker.ietf.org/doc/html/rfc6066#section-3)   where it states, "Clients, however, SHOULD
   store the SNI with the PSK to fulfill the requirements of
   [Section 4.6.1](https://datatracker.ietf.org/doc/html/rfc8446#section-4.6.1)." ?  It would be helpful if you could point out the standard you're thinking we should follow.

>     // If AllowedSANs is empty, then any SAN is allowed.

Author: @howardjohn I removed that late addition, and for this iteration, TLS will just fail if no `AllowedSANs` are specified.  In future iterations we could enhance the behavior once we come to consensus on what we want.

to Gateway cert validation, but as mentioned, TLSConnectionPolicy is used primarily to signal what the application
developer expects in the connection.  Granting this expectation to any other role would blur the lines between role
responsibilities, which violates the role-oriented design principle of Gateway API. As mentioned in Non-goal #7,
providing a mechanism for the cluster operator gateway role to override gateway to backend TLS settings is not covered
by this proposal, but should be addressed in a future update.  One idea is to use two types: ApplicationTLSConnectionPolicy,
and GatewayTLSConnectionPolicy, where the application developer is responsible for the former, the cluster operator is
responsible for the latter, and the cluster operator may configure whether certain settings may be overridden by
application developers.

Using the viewpoint of the application running inside the Pods, the TLSConnectionPolicy should contain configuration
for these use cases of the Gateway as a TLS Client:

- An explicit signal that TLS should be used by this connection.
- Which Certificate Authority (CA) to use, whether the system bundled CA, an internal CA, or self-signed.

> - TLS version to expect.  This allows implementations to introduce features to disallow or force certain TLS versions.

Reviewer 12: Isn't the TLS version negotiation sufficient ? Backend can choose min TLS version it accepts, if gateway doesn't support it - nothing will help.

Author: It doesn't seem like an overcomplicated config field.  I choose to keep it in unless there are other objections.

TLSConnectionPolicy is defined as a Direct Policy Attachment without defaults or overrides, applied to a Service that
accesses the backend in question, where the TLSConnectionPolicy resides in the same namespace as the Service it is
applied to.  The TLSConnectionPolicy and the Service must reside in the same namespace in order to prevent the
complications involved with sharing trust across namespace boundaries.  By choosing the Service resource rather than the
Route resource, we can reuse the same TLSConnectionPolicy for all the different Routes that might point to this Service.
For the use case where certificates are stored in their own namespace, users may create Secrets and use ReferenceGrants
for a TLSConnectionPolicy-to-Secret binding.  Implementations must respect a ReferenceGrant for cross-namespace Secret

> sharing to TLSConnectionPolicy, even if they don't for other cross-namespace sharing.

Reviewer 12: I'm still not sure I understand what the Secrets are doing in this - is there any real system where a server gives the client a certificate to use ? Normal authentication is that each workload ( including gateway ) has an identity that it presents - there are some mechanism in the handshake to allow some choices, but I have not seen this in many mTLS implementations.

Author: We are not addressing mTLS in this proposal.

This section discusses cross-namespace sharing, and the requirement for TLSConnectionPolicy and Service to be in the same namespace.  For the use case where certificates are stored in their own, separate namespace, users may create Secrets and use ReferenceGrants.

One or more (up to 64) subject alternative names may be declared.  SAN certificates allow authentication with a single
certificate that can serve multiple domain
names that are intended to represent the same endpoint.  For example, these five SAN hosts: example.com, 
www.example.com, example.net, IPv4 x.x.x.x, and IPv6 x.x.x.x.x.x.x.x all resolve to the same endpoint. If `AllowedSANs`
is set, then an implementation must validate that the backend pod-provided certificate represents a SAN in its list.
If empty, then TLS fails.  In the future, we will define more specific ways to handle the case of an empty `AllowedSANs`
list.

Finally, TLSConnectionPolicy allows the configuration of expected TLS versions via the `TLSVersions` field.
TLS versions specify one or more (up to 3) TLS versions that the connection may use.  If the `TLSVersions` field is set,
then the implementation must use the given versions if they are allowed.  If none of the versions are allowed, then the
TLS connection must fail.  If no TLS versions are specified, then an implementation may choose to apply a self-determined

> default version.

Reviewer 12: AFIAK TLS version is based on handshake between client and server, i.e. largest version supported by both. 

A TLSVersion config won't change the range of versions compiled in either client nor server. A backend can simply be compiled with specific version in its options. 

Not sure why the K8S Gateway needs to define alternate API to TLS.

Author: TLS version config here is optional, and the implementations may or may not use it to filters failures before they occur, or provide a default that they support if they don't support all versions.

type TLSConnectionPolicy struct {
    metav1.TypeMeta   `json:",inline"`
    metav1.ObjectMeta `json:"metadata,omitempty"`

    // Spec defines the desired state of TLSConnectionPolicy.
    Spec TLSConnectionPolicySpec `json:"spec"`

    // Status defines the current state of TLSConnectionPolicy.
    Status TLSConnectionPolicyStatus `json:"status,omitempty"`
}

// TLSConnectionPolicySpec defines the desired state of
// TLSConnectionPolicy.
// Note: there is no Override or Default policy configuration.
//
// Support: Core
type TLSConnectionPolicySpec struct {
    // TargetRef identifies an API object to apply policy to.
    // Services are the only valid API target references.
    // Note that this config applies to the entire referenced resource
    // by default, but this default may change in the future to provide
    // a more granular application of the policy.
    TargetRef gatewayv1a2.PolicyTargetReference `json:"targetRef"`

    // TLS contains TLS connection policy configuration.
    TLS *TLSConnectionPolicyConfig `json:”tls”`
}

// TLSConnectionPolicyConfig contains TLS connection policy configuration.
type TLSConnectionPolicyConfig struct {

>     // TrustedCACertRefs contains one or more references to

Reviewer 12: As discussed in the meeting, can we also add the SNI to use ? SNI is used in virtually all https connections on the internet, and allows a backend to select what cert to provide.

If we do add the SNI - could we also modify the AllowedSAN to state "if no AllowedSAN is provided, the SNI will be used to check" ( since this is also the most common scenario on the internet )

Author: From what I understand, the SNI has to be singleton on the TLSConnectionPolicy, and is sent by the Gateway in this process, in order for the backend application to choose which of several different certificates it would provide.  I don't think this is use case we need to support, as we want the TLSConnectionPolicy to be able to serve several endpoints.

i did not hear alot of support for adding SNI at the meeting.  i'll open a slack discussion and find out if there is widespread support, and whether TLSConnectionPolicy is the place to support it.

    // Kubernetes objects that contain PEM-encoded TLS certificates,
    // which are used to establish a TLS handshake between the gateway
    // and backend pod.
    //
    // If the TrustedCACertRefs is empty, then the system trusted
    // certificates should be used. If there are none, or the
    // implementation doesn't define system trusted certificates,
    // then a TLS connection must fail.
    //
    // A single TrustedCACertRef to a Kubernetes ConfigMap has "Core"
    // support.  Implementations MAY choose to support attaching
    // multiple certificates to a backend, but this behavior is
    // implementation-specific.
    //
    // References to a resource in a different namespace are 
    // invalid.
    // 
    // Support: Core - An optional single reference to a Kubernetes
    // ConfigMapName.
    //
    // Support: Implementation-specific (No reference, more than one
    // reference, or other resource types; service mesh may ignore.)
    //
    // +kubebuilder:validation:MaxItems=64
    // +optional
    TrustedCACertRefs []ConfigMapName `json:”trustedCACertRefs,omitempty”`

    // AllowedSANs are the Subject Alternative Names that are allowed for

>     // this TLS connection.  It represents the hostnames that may

Reviewer 12: or URL SANs ( not only hostnames SANs - spiffee uses URLs instead )

Author: I'm not aware of SPIFFE in any other contexts for Gateway API, so it would be great if you could open an issue on this to gauge interest instead of bringing it up here.

From what I understand, SPIFFE would want to strip the hostname from the url anyway so it could be used to validate the X.509 certificate:
https://www.envoyproxy.io/docs/envoy/latest/api-v3/extensions/transport_sockets/tls/v3/tls_spiffe_validator_config.proto

## Open Questions

This section is to record issues that were requested for discussion in the API section before this GEP moves
out of `Provisional` status.

1. Bowei recommended that we mention the approach of cross-namespace referencing between Route and Service.
Be explicit about using the standard rules with respect to attaching policies to resources.  This is mentioned in the
API section.
2. Costin recommended that Gateway SHOULD authenticate with either a JWT with audience or client cert
or some other means - so gateway added headers can be trusted, amongst other things.  This is out of scope for this
proposal, which centers around application developer persona resources such as HTTPRoute and Service.
3. Costin mentioned we need to answer the question - is configuring the connection to a backend and TLS
something the route author decides - or the backend owner?  Same for SAN (Subject Alternative Name) certificates.
The use of SAN certificates and the use of SNI, as a part of TLS, can be implementation-dependent, though the
application can still reject any request or certificate that it doesn’t support, including requests with SNI or a

> certificate with SANs.

Author: How about if I replace the last sentence with: 

"The backend owner is the application developer, and the route owner will have to collaborate with the application developer to provide the appropriate configuration for TLS.  The implementation would need to take the certificate provided by the application and verify that it satisfies the requirements of the route-as-client, including SAN information.  Sometimes the backend owner and route owner are the same entity."
